cmake/LYWrappers.cmake

```diff
@@ -90,7 +90,7 @@ define_property(TARGET PROPERTY RUNTIME_DEPENDENCIES_DEPENDS
 # \arg:AUTOGEN_RULES a set of AutoGeneration rules to be passed to the AzAutoGen expansion system
 function(ly_add_target)
 
-    set(options STATIC SHARED MODULE GEM_STATIC GEM_MODULE OBJECT HEADERONLY EXECUTABLE APPLICATION IMPORTED AUTOMOC AUTOUIC AUTORCC NO_UNITY)
+    set(options STATIC SHARED MODULE GEM_STATIC GEM_MODULE OBJECT HEADERONLY EXECUTABLE APPLICATION IMPORTED AUTOMOC AUTOUIC AUTORCC NO_UNITY EXPORT_ALL_SYMBOLS)
```

Comment:
Isn't this a potentially global change that will massively increase the size of all binaries to include all symbols?

Isn't there maybe a more local change that could be made, like, some sort of preprocessor directive or export line that can be added to just the functions in question?  The entire engine was working without this change...

Reply:
@mprettner-huawei , I'd suggest to share a before and after DLL sizes to compare.

Reply:
This change will only increase the size of shared libraries which use the `EXPORT_ALL_SYMBOLS` option, which as of this PR are `Atom_RPI.Public`, `Atom_RPI.Edit` and `Atom_RPI.TestUtils`; but will in turn decrease the size of *all* other shared libraries and executables which link to these (potential future use of this option will of course also increase the size of more shared libraries, but also decrease the size of other libraries which depend on them).

The size of these shared libraries increases somewhat when exporting all symbols (The size of the static libraries from current development is way bigger than the dlls (eg. Atom_RPI.Public.lib is 124MB))
| Library | Plain DLL | Export macros | `EXPORT_ALL_SYMBOLS` |
|---|---|---|---|
| Atom_RPI.Public.dll | 42.9MB | 44.2MB | 47.3MB |
| Atom_RPI.Edit.dll | 19.7MB | 19.8MB | 22.3MB |
| Atom_RPI.TestUtils.dll | 36.5MB | x | 36.6MB |

The previous table does not include the size of the import libraries:
| Library | Export macros | `EXPORT_ALL_SYMBOLS` |
|---|---|---|
| Atom_RPI.Public.lib | 2.37MB | 13.8MB |
| Atom_RPI.Edit.lib | 0.3MB | 7.6MB |

On the other hand, the size of some other libraries decreases (random selection of shared libraries which link to `Atom_RPI.Public`):
| Library | Before change to RPI dll | After change to RPI dll |
|---|---|---|
| EditorLib.dll | 91.5MB | 83.4MB |
| Atom_Feature_Common.Editor.dll | 50.3MB | 42.1MB |
| Multiplayer.Editor.dll | 77.2MB | 66.1MB |

For a complete picture of the build artifacts I looked at the size of each .dll and .lib file in the `bin/profile` and `lib/profile` folders with commit form before this PR and with this PR when building the "Editor" target with a clean build directory (Details: [sharedlib-comparison.xlsx](https://github.com/user-attachments/files/17204159/sharedlib-comparison.xlsx)). The summary of the excel sheet is that the build artifacts (.dll and .lib files) are ~596MB smaller (-4.9%) than before.

The reason for using the `EXPORT_ALL_SYMBOLS` version instead of export macros is mostly for convenience, since I also tried adding an export macro to all public classes, but there were some problems:
* If a base class of an exported class is itself not exported, this triggers a warning, and this warning needs to be silenced with the `AZ_PUSH_DISABLE_DLL_EXPORT_BASECLASS_WARNING` and ocasionally `AZ_PUSH_DISABLE_DLL_EXPORT_MEMBER_WARNING` macros, or more globally for the entire library; or all baseclasses need to be converted to have a dll interface themselves, which is another big undertaking
* While I could successfully build and link the `Atom_RPI.Public` and `Atom_RPI.Edit` libraries with export macros, I was not able to fix the linker errors that showed up when dependent targets try to link to these libraries in a reasonable time (errors like `lld-link: error: public: void __cdecl AZ::Internal::NonIdHandler<class AZ::SystemTickEvents, class AZ::SystemTickEvents, struct AZ::Internal::EBusContainer<class AZ::SystemTickEvents, class AZ::SystemTickEvents, 0, 1>>::BusConnect(void) was replaced`, which point to files in AzCore, so not sure how to go about this). This can probably be fixed with more effort, but I didnt want to invest even more time in the "perfect" solution if the "good enough" solution worked immediately.
* The sizes of the dlls with export macros were `Atom_RPI.Public` 44.2MB and `Atom_RPI.Edit` 19.8MB, so only a bit larger than the dlls without exporting any symbold; but since I did not manage to link to these libraries successfully, these numbers should be taken with a grain of salt.